### DISCUSSION TITLE
fwi objective dsyn process

### DISCUSSION THREAD

**Discussion by venividivici666** [slimgroup/JUDI.jl#239] [General]
when I use fwi_objective(model0, q, d_obs; misfit=misfit), how can I process the modeling forward waveform (dsyn)? I try to precess the waveform in my misfit, but I cannot use dsyn.data (type Array has no field data). 
I mean when I try “FWI with user provided misfit function”notebook, I want to process the dsyn, but fwi_objective(model0, q[i], d_obs[i]; misfit=misfit) don’t process the dsyn, and when I process the dsyn in my misfit function, there will show an error like what I have mentioned before. So I want to know whether I should modify fwi_objective function to input the dsyn after I processed. Thank you!

**Comment by mloubout** [comment]
What do you mean by "process"?

An example runnable script is always helpful for easier debug if there is an issue.

**Comment by venividivici666** [comment]
I mean if I want to mute and do bandpass or something else for dsyn, how can I input the dsyn after processed to fwi objective function to obtain call and gradient?

---- Replied Message ----
| From | Mathias ***@***.***> |
| Date | 04/02/2024 02:00 |
| To | ***@***.***> |
| Cc | venividivici666 ***@***.***>,
Author ***@***.***> |
| Subject | Re: [slimgroup/JUDI.jl] fwi objective dsyn process (Discussion #239) |

What do you mean by "process"?

An example runnable script is always helpful for easier debug if there is an issue.

**Comment by venividivici666** [comment]
By the way, when I use "fval, gradient = fwi_objective(model0, q[1], d_obs[1]; options=opt, data_precon=Ml)", there will be an error:MethodError: no method matching multi_src_fg(::JUDI.IsoModel{Float32, 2}, ::judiVector{Float32, Matrix{Float32}}, ::judiVector{Float32, Matrix{Float32}}, ::Nothing, ::JUDIOptions, ::Bool, ::Bool, ::DataMute{Float32, :turning}). Ml = judiDataMute(q.geometry, d_obs.geometry; t0=.2)

**Reply by mloubout** [reply]
Is your veersion of JUDI  > 3.4 as the `data_precon` feature is a brand new addition (last week)

**Comment by venividivici666** [comment]
How can I update JUDI to the newest version

---- Replied Message ----
| From | Mathias ***@***.***> |
| Date | 04/02/2024 23:00 |
| To | ***@***.***> |
| Cc | venividivici666 ***@***.***>,
Author ***@***.***> |
| Subject | Re: [slimgroup/JUDI.jl] fwi objective dsyn process (Discussion #239) |

Is your veersion of JUDI > 3.4 as the data_precon feature is a brand new addition (last week)

**Reply by mloubout** [reply]
just in julia

```julia
julia>  ] up JUDI
```

or on the command line `julia -e 'using Pkg;Pkg.update("JUDI")'`

**Comment by venividivici666** [comment]
I have tried to ] up JUDI, but it shows JUDI v3.3.10
